traversal/scanner/GraphIterator.java

```diff
@@ -263,34 +248,32 @@ public void recycle() {
     private class VertexTraverser {
 
         private final ProcedureVertex<?, ?> procedureVertex;
-        private final Map<ProcedureEdge<?, ?>, Vertex<?, ?>> edgeInputs;
-        private final SortedMap<Map<ProcedureEdge<?, ?>, Vertex<?, ?>>, Vertex<?, ?>> edgeInputIntersectionCache;
+        private final Scope localScope;
+        final Set<ProcedureVertex<?, ?>> implicitDependents;
+        private final Set<ProcedureVertex<?, ?>> implicitDependees;
         private Forwardable<Vertex<?, ?>, Order.Asc> iterator;
         private Vertex<?, ?> vertex;
 
         private VertexTraverser(ProcedureVertex<?, ?> procedureVertex) {
             this.procedureVertex = procedureVertex;
-            this.edgeInputs = new HashMap<>();
-            this.edgeInputIntersectionCache = new TreeMap<>(Comparator.comparing(Map::size));
-        }
-
-        private void takeNextVertex() {
-            assert hasNextVertex();
-            vertex = getIterator().next();
+            this.localScope = procedureVertex.id().isScoped() ? scopes.get(procedureVertex.id().asScoped().scope()) : null;
+            this.implicitDependents = new HashSet<>();
+            this.implicitDependees = new HashSet<>();
         }
 
-        private boolean hasNextVertex() {
-            return getIterator().hasNext();
+        private boolean findNextVertex() {
+            Forwardable<Vertex<?, ?>, Order.Asc> iterator = getIterator();
+            while (iterator.hasNext()) {
+                vertex = getIterator().next();
+                if (verifyLoops()) return true;
```

Comment:
verify loops within the `Traverser`, and only return `true` if we both find a vertex & the loop edges are satisfied